basic_autonomy/src/helper_functions.cpp

```diff
@@ -65,18 +66,21 @@ namespace waypoint_generation
     int get_nearest_index_by_downtrack(const std::vector<lanelet::BasicPoint2d>& points, const carma_wm::WorldModelConstPtr& wm, double target_downtrack)
     {
         size_t best_index = points.size() - 1;
-        for(size_t i = 0;i < points.size(); i++){
-            double downtrack = wm->routeTrackPos(points[i]).downtrack;
-            if(downtrack > target_downtrack){
-                //If value is negative, best index should be index 0
-                best_index = std::max((size_t)0, i - 1);
 
-                RCLCPP_DEBUG_STREAM(rclcpp::get_logger(BASIC_AUTONOMY_LOGGER), "get_nearest_index_by_downtrack>> Found best_idx: " << best_index<<", points[i].x(): " << points[best_index].x() << ", points[i].y(): " << points[best_index].y() << ", downtrack: "<< downtrack);
-                break;
-            }
+        // Find first point with a downtrack greater than target_downtrack
+        const auto itr = std::find_if(std::cbegin(points), std::cend(points), 
+            [&](const auto & point) { return wm->routeTrackPos(point).downtrack > target_downtrack; });
+
+        // Set best_index to the last point with a downtrack less than target_downtrack without going below index 0
+        if(itr != points.begin()){
+            best_index = std::distance(points.begin(), itr) - 1;
```

Comment:
Prefer `const` iterators over mutable ones wherever possible. Also prefer using free functions over member functions:
```cpp
if (itr != std::cbegin(points)) {
  best_index = std::distance(std::cbegin(points), itr) - 1;
} else {
  best_index = 0;
}
```

Reply:
Updated to use `const` iterators. And I might be missing something, is there an instance where I'm using a member function in this PR where a free function should be used?

Reply:
In your original implementation, you were using `points.begin()`, which is a `std::vector` member function. The free function alternative is `std::begin()`.

It's a minor detail here since we are specifically using `std::vector`, but if we had a more general function that worked on `std::vector`, `std::array`, or raw arrays, we would have to use `std::begin()` because raw arrays have no `.begin()` member function.